### ISSUE TITLE
Set default header for all requests

### ISSUE THREAD

**Issue by ghost** [a85/POSTMan-Chrome-Extension#1947]
I would like to be able to set Content-Type key to application/json value for the Header as a default, so that I do not have to set it manually for every new request.

**Comment by benmalinowski** [comment]
I second the need for this. I think there might have already been a request made for this but I'm not sure. In older versions of postman I was able to do this sort of which was one of the reasons I held out on upgrading for so long.

**Comment by numaanashraf** [comment]
@benmalinowski Did you mean header presets in the previous app?

**Comment by benmalinowski** [comment]
I mean when I go to a new item in my collection, my header was already applied to the next api call and I could make an api call without having to go to Authorization tab and selecting basic each and every time.

**Comment by benmalinowski** [comment]
This could be solved if I saved my basic auth header for every call (currently using over 500 different api calls) but I don't want to do that as I export my collection and send to other members of my team.

**Comment by cen1** [comment]
I need this to add an Authorization header with OAuth tokens to every single request. It is so damn annoying to need to copy and paste it every single time. This would be a major feature!

**Comment by easedev** [comment]
@cen1 look up presets and also environments with variables: https://www.getpostman.com/docs/environments

**Comment by AdGalesso** [comment]
@easedev even if you create a environment variable you have to go to each request and add the headers =S. It would be great if you could bulk update all your requests.

**Comment by robertjd** [comment]
+1 for the ability to have a preset always applied to new requests.  Perhaps this is bound to an environment, rather than global?

**Comment by a85** [comment]
Folks, this will be available through the Collections feature instead through environments in the future. You will be able to set collection level auth and scripts that will allow for all sorts of use cases. Closing this particular feature request for now.

**Comment by nathaniel-holder** [comment]
+1

**Comment by anudhingra** [comment]
+1

**Comment by marcioAlmada** [comment]
+1

**Comment by kamilwlf** [comment]
+1

**Comment by mattcrwi** [comment]
+1

**Comment by rbluethl** [comment]
+1

**Comment by ctmlr** [comment]
+1, this was marked as closed but I don't see any way to accomplish this still.

**Comment by dlgoodchild** [comment]
+1

**Comment by yarandoo** [comment]
+1

**Comment by Ariex** [comment]
+1

**Comment by dnipser** [comment]
+1

**Comment by fbatiga** [comment]
+1

**Comment by sebastian-serfe** [comment]
+1

**Comment by DannyFeliz** [comment]
+1

**Comment by ahghatol** [comment]
+1

**Comment by jusdeleon** [comment]
+1

**Comment by Free1man** [comment]
+1

**Comment by khanhtdbse** [comment]
+1

**Comment by 2bj** [comment]
+1

**Comment by cen1** [comment]
I guess this is kinda solved now by using global or env variables in postman collections. For example, adding a JWT token to all requests in a collection:
1. Acquire the auth token and write a test that sets a global variable.
2. Use the global variable in Authentication header in other collection requests.

Not 100% what people here have in mind but shuld have the same effect on collection level.

**Comment by mpereira-dev** [comment]
@cen1 that is a work around but it requires editing every request manually. There isn't an easy way that I know of to add a header that is applied to all / a larger subset of requests. (Like apply / remove a preset to an entire folder or collection) @a85 any progress on this feature, no idea why you closed it. Similar issue: https://github.com/postmanlabs/postman-app-support/issues/1575

**Comment by dlgoodchild** [comment]
Yeah this work around is pretty laborious, I really wish they'd consider a proper solution and reopen this.

**Comment by mpereira-dev** [comment]
If you can't wait here is a work around I just made:

1. Export your collection (data format v2.1)
2. Open firefox , dev tools, scratch pad
3. Replace the header information with your header
4. Paste the code below replacing the var a with your contents of the exported .json file
5. Run the script
6. The copy(b) command will put the new data with in your clipboard
7. In postman, click import > Paste Raw Text > Import > as a copy.
8. Verify your requests have your header, and run it :)

```
var myHeader = {
  "key": "X-Client-DN",
  "value": "{{Postman-DN}}",
  "description": "The User's DN Interacting with the system."
};

function addHeader(obj,header){
  if(obj.hasOwnProperty('request')){
    obj.request.header.push(myHeader)
  }
  if(obj.hasOwnProperty('item')){
    obj.item.forEach(function(element){
      element = addHeader(element,header);
    });
  }
  return obj;
}
```

`var a = paste your postman collection (v2.1) export output;`

`var b = addHeader(a,myHeader);`

`copy(b);`

**Comment by cen1** [comment]
@mpereira-dev not sure what you mean with "editing every request manually". You simply add for example: `Authorization Bearer {{jwtToken}}` in every collection request and it will auto-fill. So once you have the collection set up with variables there is zero editing involved. You can then either set the variable automatically when executing the first auth request or set it once in the env json file and it propagates everywhere. 

Unless you mean adding an additional header? But I don't really see a use case for that. You usually know in advance which set of headers apply for each request.

**Comment by mpereira-dev** [comment]
@cen1 Yeah the use case is for a new header. I love postman and environment variables are great! But in my case the security requirements (or my assumption of the requirements) changed after I built a large collection. So needed to go back and add a new header to each request. Also in the near future I will need to go back and modify a subset of the requests by either replacing the header or adding a new one. Currently there is no functionality that allows you to edit multiple requests simultaneously or have a sort of parent child relationship where the configuration or tests etc are inherited and managed in one place. We got save as and duplicate which is a start!

**Comment by xavier-b** [comment]
+1

**Comment by richp10** [comment]
+1 and I am not sure this is being properly understood.  The issue is that if you have big api and for example change from basic-auth to jwt, you need to go through every endpoint adding the Auth bearer header - using {{token}} tags helps, but it is still painful.

**Comment by ramseyjacob** [comment]
@richp10 - Exactly!  Having Collection level injection of headers before the request is sent and allow the value of the injected header to be pulled from an environment variable...would be a perfect feature.

**Comment by taotao365s** [comment]
+1

**Comment by tylrd** [comment]
With 5.4.1, we can now add variables, authorization, pre-request and test scripts to a collection or folder, but still no default collection/folder level headers. Please add this feature - without default headers a lot of the collection-level variables and pre-request scripts are not useful to me.

**Comment by prashant-pokhriyal** [comment]
Suppose you add a folder to a collection. Under the Authorization tab, the default authorization type is set to **Inherit auth from parent**.

The **Inherit auth from parent** setting indicates that every request in this folder by default uses the authorization type from the parent. In the following example, the collection is using **No Auth**, so the folder uses **No Auth**, meaning all requests in that folder will use **No Auth** .
![](https://s3.amazonaws.com/postman-static-getpostman-com/postman-docs/auth-folder.png)

What if you want to leave the parent collection authorization type as **No Auth**, but update this specific folder’s authorization helper? You can edit the folder details, select **Basic Auth** from the TYPE dropdown, and input your credentials. As a result, every request in this folder relies on **Basic Auth** while the rest of the requests in the parent collection still do not use any authorization.

![](https://s3.amazonaws.com/postman-static-getpostman-com/postman-docs/auth-folder-basic.png)

Similarly, if you want to update the authorization for a single request in this folder, you can simply select a different authorization type for that request.

Source: [Postman Authorization Docs](https://www.getpostman.com/docs/postman/sending_api_requests/authorization)

**Comment by Gosherm** [comment]
This doesn't work for situations where dev has implemented a non-standard bearer request. In my case, the request is Authorization = {{key}}, NOT Authorization = bearer {{key}}. Really need a way to address this at the global level, not by having to set an individual key on every single request.

**Comment by degraaf** [comment]
This should be re-opened and support added for headers. I currently use two layers of authentication - basic auth (stage environment) and api key (for the application). Without folder level headers this cannot be managed centrally. Should be manageable at folder level.

**Comment by hainee** [comment]
+1000000

**Comment by javabrett** [comment]
@degraaf and others arriving here - this issue is closed.  I've opened #4044 to push for generic Headers as a new Auth type. Might be worth adding thoughts or a thumbs-up there.

**Comment by Sparshith** [comment]
+1

**Comment by jakwo** [comment]
+1

**Comment by JKalash** [comment]
+1

**Comment by sobkowicz** [comment]
+1

**Comment by chris-burkhardt** [comment]
+1

**Comment by hashunst** [comment]
+ 1

**Comment by ssparasuram** [comment]
+1

**Comment by govindrai** [comment]
+1

**Comment by niksteff** [comment]
+1

**Comment by ghost** [comment]
+1

**Comment by fredeil** [comment]
This is a must have.

**Comment by ir-fuel** [comment]
+1

**Comment by andi-livn** [comment]
+1 indeed. I've come across several cases of APIs using a custom header for authorisation

**Comment by ma-schmidt-de** [comment]
+1. The use case of having a default header "origin=XX" or "apikey=XX" is super common. There must be a central place to define headers for Collections and Folders.

**Comment by Pedr0Rocha** [comment]
+1

**Comment by mmieluch** [comment]
+1

**Comment by Geolim4** [comment]
+1

**Comment by Geolim4** [comment]
@postman-admin Can you please re-consider adding this feature ?

**Comment by manuelottlik** [comment]
+1

**Comment by govindrai** [comment]
@a85 since collection level auth is already out, is this a feature still in the works? I only ask because this ticket is closed and the feature was slated to be released as per your response in Aug 2017. Cheers.

**Comment by DannyFeliz** [comment]
I really need this feature, I use Postman a lot daily

**Comment by tmbdrogba** [comment]
+1

**Comment by afeblot** [comment]
+1

**Comment by rbrunge** [comment]
+1

**Comment by zero-master** [comment]
TLDR; for a project manager

Add a way to define custom authorization routine for the user. It will help many of us who are using nonstandard way of authorization as we are working with legacy applications.

https://www.getpostman.com/docs/v6/postman/sending_api_requests/authorization

It will solve this issue.

**Comment by mmieluch** [comment]
It's not just for custom auth. We're working on an API built with Dingo on top of Laravel. Dingo uses the "Accept" header to determine the version of the API the request is supposed to be routed to. Currently we have to add the header manually to every single request. We store the header content in a variable, but adding the header is yet another thing to remember when creating new requests in Postman.

**Comment by guillermoasaro** [comment]
+1
Same as others. It is a pain to test an Ruby on Rails API with Device auth

**Comment by ADTC** [comment]
+1 (these just keep rolling here! why isn't this getting any attention???)

In my case, I was stumped because the API devs made it to use a custom authorization. Instead of the standard `authorization: Bearer <token>`, it is a non-standard version `access_token: <token>`. Postman currently provides no way to automate this non-standard authorization because there's no "Custom Auth" option where I could just define it as being a simple parameter, request header or request body to fit my unique needs.

I'd love to see this:
![mock-up of custom auth option in Edit Collection Authorization tab](https://user-images.githubusercontent.com/6047296/37234361-4bcd00de-241d-11e8-8a63-2c20756549d9.png)

**Comment by kobelobster** [comment]
+1

**Comment by haf** [comment]
1. Export collection as v2.1 schema
 2. Create file:

```node
#!/usr/bin/env node
'use strict';

var hostHeader = {
  "key": "Host",
  "value": "{{hostname}}",
  "description": "The spoofed hostname, to avoid having to set the domain in /etc/hosts"
};

function addHeader(obj, header) {
  if (obj.hasOwnProperty('request')) {
    obj.request.header.push(header);
  }

  if (obj.hasOwnProperty('item')) {
    obj.item.forEach(function(element) {
      element = addHeader(element, header);
    });
  }

  return obj;
}

const fs = require('fs');
const input = fs.readFileSync(0); # stdin
const collection = JSON.parse(input);
const result = addHeader(collection, hostHeader);
process.stdout.write(JSON.stringify(result));
```

 3. `chmod +x add-header.js`
 4. `cat my-collection.json | ./add-header.js | jq . >my-collection.v2.json`
 5. Import collection

**Comment by nikspatel007** [comment]
I would love default headers option as well. For now, I have been using `Presets` feature to add multiple headers to the new requests as I create them.

https://www.getpostman.com/docs/v6/postman/sending_api_requests/requests &rarr; `Header Presets`

**Comment by zeeshanu** [comment]
+1

**Comment by aligorji** [comment]
+1

**Comment by MMasterson** [comment]
+1

**Comment by jonikl** [comment]
+1

**Comment by AImager** [comment]
+1

**Comment by prashant-pokhriyal** [comment]
Please, guys, it's my humble request not to spam here by just commenting **+1**. GitHub has given you emojis to express your emotions. You can make use of thumbs up if you are in motion of the OP. You can see for yourself how awkward it is looking while going over the thread. If possible delete your comments which contains only **+1**.

**Comment by kobelobster** [comment]
@prashant-pokhriyal Normally, I'd agree with you, however this issue exists for almost two years already and neither a contributor, nor a maintainer or owner has EVER written a comment in here. This issue doesn't get any attention anyway.

**Comment by ADTC** [comment]
@prashant-pokhriyal can you tell me if you got an email notification or GitHub notification for my confused emoticon on your comment? If not, that's the problem with just adding emoticons into the post or existing comments. No one gets notified. At least with an actual comment everyone gets an email unless they turned emails off.  Or they get a notification inside GitHub.

Also, project owners and moderators can delete all "+1 only" comments if they want to. Perhaps they'd want to do that and add the names of +1ers to the issue's top post.

**Comment by govindrai** [comment]
+1 for collection level headers.

**Comment by rhysv96** [comment]
+1, please add this

**Comment by EvilVir** [comment]
+1 Or at least please add "Custom" authorization setting so we can set custom value for that header.

**Comment by dolmen** [comment]
@ADTC There is a ["subscribe" button](https://help.github.com/articles/subscribing-to-and-unsubscribing-from-notifications/) on the right side that does what you need without spamming everyone who is already subscribed.

So, please:
- use [GitHub reactions](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)
- use the [Subscribe button](https://help.github.com/articles/subscribing-to-and-unsubscribing-from-notifications/)
- delete your uselesss "+1"
- don't add "+1" that spam the 74 users already subscribed

**Comment by barakbd** [comment]
I am trying to set a header in a pre-request script.
This way:
```
pm.request.headers.add({key:"value"})
pm.request.headers.add("key", "value")
```
and this way:
```
pm.request.headers.append({key:"value"})
pm.request.headers.append("key", "value")

```
Both not working.
What am I doing wrong?

**Comment by andi-livn** [comment]
Hi Barak.

There appears to be no way to directly add or modify headers from the pre-request scripts, which would allow doing what has been requested here via the collection level pre-request script.

Unfortunately the only workaround I found is to define a global variable, e.g. `myKey`, set a header on the request with that variable as value e.g. `key:{{myKey}}` (both done via the Postman GUI) and to then set the value of that global variable in the pre-request query, e.g. `pm.globals.set("myKey", value);`.

**Comment by phemonick** [comment]
```
const setAuthorizationToken = (token) => {
  if (token) {
    axios.defaults.headers.common.authorization = `${token}`;
    console.log({
      token,
      auth: 'authorization set',
    });
  } else {
    delete axios.defaults.headers.common.authorization;
  }
};

``` when using axios, u pass in your token as a parameter

**Comment by potsed** [comment]
So I would love this too.

## User Story
As a DEVELOPER, I want to SET A CUSTOM HEADER in Postman for all requests AT A COLLECTION AND FOLDER LEVEL, so that I can test the api in different ways. 

## Scenarios
Scenario: Add header to all requests
Given the collection has > 1 request
When a Postman Collection has Default Collection Headers
Then the Default Collection Headers should be sent with every request from the Collection
Where they are not overwritten by Folder Headers and/or Request Headers

Scenario: Add header to some requests
Given a folder has > 1 request
When the Postman Folder has Default Folder Headers
Then the Default Folder Headers should be sent with every request in the Folder
Where they are not overwritten by Request Headers

## Example: 
### Simulating XMLHttpRequests
```
Current options: 
  - Add header to every request (Sometimes 100's or 1000's)
  - Workarounds using the env variables and pre-request scripts
  - Header presets, useful but:
      - Still required to add the preset to each request (Sometimes 100's or 1000's)
      - Updating preset does not update existing requests using the preset
      - Weird issue where it adds Authorize settings even though they are not defined

Process could be: 
  - Add header to collection: X-Requested-With: XMLHttpRequest
```

**Comment by StephaneJuban** [comment]
+1 😜

**Comment by leonstel** [comment]
+1

**Comment by Bheemaanil** [comment]
any update on this?

**Comment by DannyFeliz** [comment]
I ended up using [insomnia](https://insomnia.rest) instead, allow me to do this and is pretty extensible and allow me to import my postman collections. Very recommended.

**Comment by muhammadbahrami** [comment]
please add

**Comment by razorsharpshady** [comment]
Was looking to add custom header for collection, but it looks like the postman team isn't really in support of this feature, otherwise it's been 2 years since this issue was first reported they could've added this feature.

**Comment by mateusfccp** [comment]
I confirm the need of this feature.

**Comment by J5Dev** [comment]
This so needs to be added... over 800 api requests in a collection, and due to changes in our app, I now need to be explicit about the content type header being sent.

I now have to manually update each one, adding the headers... even though every single one is teh exact same!

_*Disclaimer, I know there are semi-programatic ways... I can export, find/replace, but it isn't the point._

This feature is exceptionally needed

**Comment by andi-livn** [comment]
> 
> 
> ... over 800 api requests in a collection, and due to changes in our app, I now need to be explicit about the content type header being sent.
> 
> I now have to manually update each one, adding the headers... even though every single one is teh exact same!

Ouch @J5Dev! I agree this is a much needed feature. You might be better off exporting your collection to JSON, modifying or patching in those content type headers and importing it back into postman. Beats clicking about in the GUI for a fortnight.

**Comment by J5Dev** [comment]
> > ... over 800 api requests in a collection, and due to changes in our app, I now need to be explicit about the content type header being sent.
> > I now have to manually update each one, adding the headers... even though every single one is teh exact same!
> 
> Ouch @J5Dev! I agree this is a much needed feature. You might be better off exporting your collection to JSON, modifying or patching in those content type headers and importing it back into postman. Beats clicking about in the GUI for a fortnight.

Exactly what I did... But, as said it isn't the point. It would have still been faster in the gui to just quickly add 2 headers to the collection which applied to all calls within it.

**Comment by shamasis** [comment]
I have a thought for a quick relief from this pain, can we share feedback? Here it goes:

We introduced folder and collection level scripts that adds the ability to attach a test/pre-req script at folder / collection and that gets run every time a request gets run. Now, if we add the ability to modify the request from pre-request script using `pm.request.*` APIs.

This would alleviate the problem to some extent, at least in a programmatic way. This would also open up new ways of doing stuff with the request now being editable from script. While this is released, we can work on two other possible solutions to this: (1) introduce adding request level template-y stuff at folder/collection or (2) add bulk find and replace / find and append in app.

**Comment by barakbd-bluevine** [comment]
@shamasis - have you tried it? I tried it a while ago, and I ran into this issue - request object is read only:

> request {object}: Postman makes the request object available to you while writing scripts. This object is read-only. Changing properties of this object will have no effect. Note: Variables will NOT be resolved in the request object. The request object is composed of the following:

https://learning.getpostman.com/docs/postman/scripts/postman_sandbox#requestresponse-related-properties

**Comment by shamasis** [comment]
@barakbd - I am suggesting this. It is not yet done that way. What are your thoughts?

**Comment by jhardin-aptos** [comment]
It would be nice to not have to drop into code to do this.

I'd like to see a "Headers" tab in the Edit Collection and Edit Folder UIs... Define a header there and it's included in all requests in that collection/folder. Individual folders/requests could override the collection/folder default header by redefining the header explicitly in that one folder/request.

**Comment by wabiloo** [comment]
There seems to be wide agreement that this is needed. How much would it take for the Postman team to accept the "will of the people"?  (guess what country I live in...)

**Comment by jguyenot** [comment]
I confirm the need of this feature.

**Comment by OzanKurt** [comment]
Seriously,

There is no update about this problem for 2 years?

WAKE UP!

**Comment by ir-fuel** [comment]
Who should wake up?
How much did you pay for Postman?

**Comment by E1337Kat** [comment]
I mean, if the feature was added as a postman pro feature, I would absolutely have my company get it for us. This is one of the most requested features from what I have seen, and I can't see how it would be too difficult to implement injecting headers from the folder/collection. Not to downplay the amount of work it would take to implement of course.

**Comment by chengchengmu** [comment]
It looks like a good soul : @codenirvana  implemented the feature here https://github.com/postmanlabs/postman-runtime/pull/735

**Comment by codenirvana** [comment]
As mentioned by @shamasis https://github.com/postmanlabs/postman-app-support/issues/1947#issuecomment-457740023

We've added support for modify the request from pre-request script on our Canary (`version: 7.1.0-canary01`) channel https://www.getpostman.com/canary

Using `pm.request.headers.*` API in Collection's (or Folder) Pre-Request script you can add, remove or update headers for every request in that collection.

**Setting a default header for all requests**: 

<img width="811" alt="Screenshot 2019-03-29 at 2 32 34 AM" src="https://user-images.githubusercontent.com/4865836/55192637-efcddb00-51ca-11e9-9d59-1e40678ae7d3.png">

> This will add `Content-Type` header for every request in `Test` collection. 

Following will work in `pre-request` scripts:
```javascript
// Add new header
pm.request.headers.add({
    key: 'Accept-Encoding',
    value: 'gzip'
});

// Add or update an existing header
pm.request.headers.upsert({
    key: 'Connection',
    value: 'close'
});

// remove header
pm.request.headers.remove('User-Agent')
```
Please try it out in the Canary version and let us know if you face any issue.
Will update the thread once we release this on our stable channel.

**Comment by codenirvana** [comment]
This has been added in the latest Postman App (v7.0.9). 🎉
Refer above comment for the detailed usage.

**Comment by dandelaney8** [comment]
Are there any plans to add this on a global level? For example, I'm trying to set some headers for every single request across multiple collections, and it's annoying to have to maintain the same code in every single one.

**Comment by khushboobansal1910** [comment]
> As mentioned by @shamasis [#1947 (comment)](https://github.com/postmanlabs/postman-app-support/issues/1947#issuecomment-457740023)
> 
> We've added support for modify the request from pre-request script on our Canary (`version: 7.1.0-canary01`) channel https://www.getpostman.com/canary
> 
> Using `pm.request.headers.*` API in Collection's (or Folder) Pre-Request script you can add, remove or update headers for every request in that collection.
> 
> **Setting a default header for all requests**:
> 
> <img alt="Screenshot 2019-03-29 at 2 32 34 AM" width="811" src="https://user-images.githubusercontent.com/4865836/55192637-efcddb00-51ca-11e9-9d59-1e40678ae7d3.png">
> 
> > This will add `Content-Type` header for every request in `Test` collection.
> 
> Following will work in `pre-request` scripts:
> 
> ```js
> // Add new header
> pm.request.headers.add({
>     key: 'Accept-Encoding',
>     value: 'gzip'
> });
> 
> // Add or update an existing header
> pm.request.headers.upsert({
>     key: 'Connection',
>     value: 'close'
> });
> 
> // remove header
> pm.request.headers.remove('User-Agent')
> ```
> 
> Please try it out in the Canary version and let us know if you face any issue.
> Will update the thread once we release this on our stable channel.

upsert is not working. It is appending the new value to previous value and not updating it to a new value.

**Comment by sharunkumar** [comment]
> upsert is not working. It is appending the new value to previous value and not updating it to a new value.

upsert is still not working the way that its mentioned 🤔

**Comment by shubhbhargav** [comment]
@sharunkumar Can you please share a screen recording of the issue to help us debug this further? I tried it out on the latest version (9.22.5-220624-0940) and could see that the upsert is working fine for me. Sharing the screen recording below:

https://user-images.githubusercontent.com/41579749/175827927-f6048810-7f47-4ebd-aead-b1d0c4ec2c92.mov
